### PR TITLE
Update schema to merge changes made across OASIS and SARIF-SDK versions

### PR DESCRIPTION
**Pull request by michaelcfanning**
Sync schema to converge (useful) edits that materialized in the SARIF SDK repository. @lgolding

### CODE REVIEW

File: Schemata/sarif-schema.json [hunk 1]
```diff
@@ -353,6 +272,54 @@
       }
     },
 
+    "externalFiles": {
```
Comment (michaelcfanning) [inline Schemata/sarif-schema.json]: >externalFiles [](start = 5, length = 13)

@lgolding, please review this definition carefully. i know you looked at it in context of the SARIF SDK repo, but in isolation here perhaps other improvements occur to you,

File: Schemata/sarif-schema.json [hunk 2]
```diff
@@ -1270,6 +1237,11 @@
           "$ref": "#/definitions/message"
         },
 
+        "analysisTarget": {
```
Comment (michaelcfanning) [inline Schemata/sarif-schema.json]: >analysisTarget [](start = 9, length = 14)

FYI, moving items around in the schema actually has impact on the generated OM. The class definition code emit follows the schema. This ordering appears to have some influence on compilation/MSIL code gen (to use the C# case). Reflection adheres to this ordering. As a result, changes in schema ordering can result in changes to JSON data ordering as emitted by the C# object model. if we want to define a standard for ordering (such as everything is alphabetized at its property scope), we should do that and make a scrub.

File: Schemata/sarif-schema.json (hunk 1 above)
Comment (lgolding) [reply Schemata/sarif-schema.json]: Huh. This isn't what I had in mind but I think it's better.

I wasn't going to specify the property names that could appear; I was just going to allow arbitrary property names, and each property value would be an array of `fileLocation`s. And there is text in the change draft that says "that array can have one or more elements if the externalized property is an array, otherwise it can have only one element."

But by explicitly enumerating the property names, you've made those one-element arrays unnecessary.

So yes, this is good, and I'm going to revise my change draft to match.

---
In reply to: [215360196](https://github.com/oasis-tcs/sarif-spec/pull/236#discussion_r215360196) [](ancestors = 215360196)

File: Schemata/sarif-schema.json [hunk 3]
```diff
@@ -783,7 +750,7 @@
 
         "workingDirectory": {
           "description": "The working directory for the analysis rool run.",
```
Comment (lgolding) [inline Schemata/sarif-schema.json]: >rool [](start = 65, length = 4)

Hah, while we're at it, `s/rool/tool/g`.

**Comment by lgolding** [comment]
>           "description": "A stable, unique identifier for class of related runs to which this run belongs, in the form of a GUID.",

"class" => "the class"

---
Refers to: Schemata/sarif-schema.json:1618 in 10c1ac8. [](commit_id = 10c1ac830fba2f904a71423a4731c5ff3e23a325, deletion_comment = False)

**Comment by lgolding** [comment]
LGTM, just a couple of comment cleanup nits.

File: Schemata/sarif-schema.json [hunk 4]
```diff
@@ -353,6 +272,54 @@
...
+          "$ref": "#/definitions/fileLocation"
+        },
+
+        "graphs": {
+          "description": "An external reference to a run.graphs object to be merged with the root log file.",
+          "$ref": "#/definitions/fileLocation"
+        },
+
+        "invocations": {
+          "description": "An external reference to an run.array of invocation objects to be merged with the root log file.",
```
Comment (lgolding) [inline Schemata/sarif-schema.json]: >run. [](start = 54, length = 4)

"an run.array" => "an array"

File: Schemata/sarif-schema.json [hunk 5]
```diff
@@ -353,6 +272,54 @@
...
+          "$ref": "#/definitions/fileLocation"
```
Comment (lgolding) [inline Schemata/sarif-schema.json]: >"$ref": "#/definitions/fileLocation" [](start = 10, length = 36)

... and since the `run.invocations` property _is_ an array, its value must be an array of `fileLocation` objects, just as for `run.results`. I hadn't noticed this before either; I thought `results` was the only array-valued externalizable property.

File: Schemata/sarif-schema.json [hunk 6]
```diff
@@ -353,6 +272,54 @@
...
+        },
+
+        "files": {
+          "description": "An external reference to a run.files object to be merged with the root log file.",
+          "$ref": "#/definitions/fileLocation"
+        },
+
+        "graphs": {
+          "description": "An external reference to a run.graphs object to be merged with the root log file.",
+          "$ref": "#/definitions/fileLocation"
```
Comment (lgolding) [inline Schemata/sarif-schema.json]: >"$ref": "#/definitions/fileLocation" [](start = 10, length = 36)

`run.graphs` is actually an array of `graph` objects. So the value of this property must be an array of `fileLocation` objects.

BUT! I just realized that there is a bug in the spec. `run.graphs` (and `result.graphs`, for that matter) should _not_ be an array. It should be a dictionary whose key is `graph.id` (just like in the `resources.rules` dictionary the key is `rule.id`).

This matters because we have `graphTraversal` objects which refer to the `graph` they traverse by specifying the graph's id. So we should be able to look up the referenced graph in a dictionary. I will file a bug.

Bottom line: Leave this property alone (_don't_ make it an array) because we'll fix up the `run.graphs` property (and `result.graphs`) to be a dictionary instead of an array.

Comment (lgolding) [reply Schemata/sarif-schema.json]: I filed [Issue #237]([https://github.com/oasis-tcs/sarif-spec/issues/237](https://github.com/oasis-tcs/sarif-spec/issues/237), ), "run.graphs and result.graphs should be dictionaries, not arrays."

---
In reply to: [215688773](https://github.com/oasis-tcs/sarif-spec/pull/236#discussion_r215688773) [](ancestors = 215688773)

File: Schemata/sarif-schema.json [hunk 7]
```diff
@@ -353,6 +272,54 @@
...
     },
 
+    "externalFiles": {
+      "description": "References to external files that should be inlined with the content of a root log file.",
+      "additionalProperties": false,
+      "type": "object",
+      "properties": {
+
+        "conversion": {
+          "description": "An external reference to a run.conversion object to be merged with the root log file.",
```
Comment (lgolding) [inline Schemata/sarif-schema.json]: >An external reference to [](start = 26, length = 24)

Consider: "The location of a file containing..."

File: Schemata/sarif-schema.json (hunk 6 above)
Comment (lgolding) [reply Schemata/sarif-schema.json]: I'm sorry, I now regret my "bottom line" comment above. As the spec stands, `run.graphs` is an array, and so I revert to my previous position: The value of this property must be an array of `fileLocation` objects. On the day we fix #237, we will change this as well.

---
In reply to: [215688773](https://github.com/oasis-tcs/sarif-spec/pull/236#discussion_r215688773) [](ancestors = 215688773)

File: Schemata/sarif-schema.json (hunk 3 above)
Comment (michaelcfanning) [reply Schemata/sarif-schema.json]: nice catch. ran content through spell and grammar check and found a couple other double spaces and term simplifications.

---
In reply to: [215672927](https://github.com/oasis-tcs/sarif-spec/pull/236#discussion_r215672927) [](ancestors = 215672927)

File: Schemata/sarif-schema.json (hunk 5 above)
Comment (michaelcfanning) [reply Schemata/sarif-schema.json]: right!

---
In reply to: [215686784](https://github.com/oasis-tcs/sarif-spec/pull/236#discussion_r215686784) [](ancestors = 215686784)

**Comment by michaelcfanning** [comment]
>           "description": "A stable, unique identifier for class of related runs to which this run belongs, in the form of a GUID.",

nice

---
In reply to: [419140212](https://github.com/oasis-tcs/sarif-spec/pull/236#issuecomment-419140212) [](ancestors = 419140212)

---
Refers to: Schemata/sarif-schema.json:1618 in 10c1ac8. [](commit_id = 10c1ac830fba2f904a71423a4731c5ff3e23a325, deletion_comment = False)

File: Schemata/sarif-schema.json (hunk 5 above)
Comment (lgolding) [reply Schemata/sarif-schema.json]: This one isn't resolved; it still specifies a single `fileLocation` instead of an array of `fileLocation`s. Everything else looks good, and I agree with "Pending" on the "graphs" problem.

---
In reply to: [215779432](https://github.com/oasis-tcs/sarif-spec/pull/236#discussion_r215779432) [](ancestors = 215779432,215686784)

Comment (lgolding) [reply Schemata/sarif-schema.json]: As we discussed:
1. An `invocation` might be large because it contains `toolNotification`.
2. Each of the external files for any array-valued property contains an _array_, and it's the union of those arrays that is merged with the root file. So if all the objects are small, you can put them in one external file.

---
In reply to: [215783343](https://github.com/oasis-tcs/sarif-spec/pull/236#discussion_r215783343) [](ancestors = 215783343,215780861,215779432,215686784)
